Bluetooth: Fix buffer allocation when under syswq

Vudentz opened the pull request:
With the changes to make move L2CAP channels to be processed under syswq and the addition of checks not to block with K_FOREVER it is now possible that bt_conn_create_pdu would return NULL which has to handled by the caller. In addition to that processing under syswq would not pend waiting for a buffer when using K_FOREVER so this introduces alternative APIs to use so protocols such as SMP can pass its own timeout.

Fixes #18705

Vudentz committed:
Bluetooth: conn: Add bt_conn_create_pdu_timeout

This adds bt_conn_create_pdu_timeout function which can be used
to provide a timeout when allocating a buffer.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: SMP: Fix invalid access if a buffer cannot be allocated

In certain cases there could be no buffer available which may lead to
NULL being returned by bt_l2cap_create_pdu so instead use
bt_l2cap_create_pdu_timeout to wait with a specific timeout.

Note that this should also ensure the SMP will not be waiting for a
buffer when the remote might have timed out already.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: L2CAP: Fix invalid access if a buffer cannot be allocated

Since bt_l2cap_create_pdu can return NULL when used under syswq context
the code should always check its result, this also changes the timeout
to have a specific value (RTX maximum timeout) so signalling PDUs still
wait for a buffer to become available.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: ATT: Use timeout when allocating a buffer

Whenever responding or confirming use the ATT timeout to allocate a
buffer.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

carlescufi reviewed (changes requested):
This doesn't seem to fix it, I still get a disconnection from the phone instead of a successful pairing procedure

Vudentz committed:
Bluetooth: Don't run SMP callbacks on syswq

SMP shall be considered internal TX notify callbacks so they are
allowed to be run from TX thread context like the others.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

joerchan commented:
Should you add fixes: https://github.com/zephyrproject-rtos/zephyr/pull/18704 ?

Vudentz commented:
> Should you add fixes: #18704 ?

That is not an issue though it is the PR itself, or did you mean to past something else?

Vudentz committed:
Bluetooth: ATT: Fix att_op_get_type when client is disabled

When CONFIG_BT_GATT_CLIENT is not set att_op_get_type would return
unkown operation instead of properly return the opcode.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

carlescufi commented:
> > Should you add fixes: #18704 ?
> 
> That is not an issue though it is the PR itself, or did you mean to past something else?

Edited it myself with the correct issue number.

carlescufi reviewed (approved):
Fixes the issue, I can notify at full speed and bond now

Vudentz committed:
Bluetooth: ATT: Fix misleading warning

When an opcode doesn't have a handle that doesn't mean it is unknown
just that it will not be handle as it could have been disabled.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

joerchan commented on subsys/bluetooth/host/att.c:
```diff
@@ -1942,6 +1942,28 @@ static att_type_t att_op_get_type(u8_t op)
 		return ATT_COMMAND;
 	}
 
+#if !defined(CONFIG_BT_GATT_CLIENT)
```
Complete the list and just get rid of the rest of the function?

Vudentz replied:
Yep, Ive been thinking about that, though that means one has to remember to add new opcodes there as well but I guess it is faster to use the switch statement as well.

joerchan replied:
I think the switch statement is better, it's more robust to ifdefs in the handlers array.

ioannisg commented:
We need a bit of more approvals, here.

joerchan reviewed (commented):
There are a few places in smp.c where the return value when we don't buffers are not checked.
`smp_send_pairing_random`
`sc_smp_send_dhkey_check`
`smp_send_pairing_confirm`

`bt_smp_distribute_keys` This one returns many times, in buffer error failure, in this case there is no error returned. 

An error returned will result in an SMP Error packet being sent, with reason Unspecified.
Is it so that failing to allocate SMP buffer will always be either allocated, or wait until SMP timeout.
So then all of these errors should be handled, but they are not all handled in the same way?

Vudentz commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -415,6 +423,9 @@ static int l2cap_le_conn_req(struct bt_l2cap_le_chan *ch)
 	buf = l2cap_create_le_sig_pdu(NULL, BT_L2CAP_LE_CONN_REQ,
 				      ch->chan.ident, sizeof(*req));
+	if (!buf) {
+		return -ENOMEM;
```
I usually use ENOMEM for that, not sure if changing would make much difference to be honest.

joerchan replied:
Right, I see that this module use ENOMEM consistently. So ignore that comment.

Vudentz commented:
All the calls to smp_creat_pdu do check for NULL return:

smp_send_pairing_random:
	rsp_buf = smp_create_pdu(conn, BT_SMP_CMD_PAIRING_RANDOM, sizeof(*req));
	if (!rsp_buf) {
		return BT_SMP_ERR_UNSPECIFIED;
	}

sc_smp_send_dhkey_check:
	buf = smp_create_pdu(smp->chan.chan.conn, BT_SMP_DHKEY_CHECK,
			     sizeof(*req));
	if (!buf) {
		return BT_SMP_ERR_UNSPECIFIED;
	}

smp_send_pairing_confirm:
	buf = smp_create_pdu(conn, BT_SMP_CMD_PAIRING_CONFIRM, sizeof(*req));
	if (!buf) {
		return BT_SMP_ERR_UNSPECIFIED;
	}

Not sure what you are refering to @joerchan, we may have to return a different error if we can't get buffer allocated in time but that is a different matter.

joerchan commented:
@vudentz I mean that the return code of smp_send_pairing_confirm, and the other ones I mentioned, is ignored

joerchan commented on subsys/bluetooth/host/smp.c:
```diff
@@ -1949,7 +1954,7 @@ static void bt_smp_distribute_keys(struct bt_smp *smp)
 		}
 
-		smp_send(smp, buf, sign_info_sent, NULL);
+		smp_send(smp, buf, smp_sign_info_sent, NULL);
```
Need same change in `smp_br_distribute_keys`

joerchan commented on subsys/bluetooth/host/smp.c:
```diff
@@ -3251,7 +3272,7 @@ static u8_t smp_pairing_random(struct bt_smp *smp, struct net_buf *buf)
+	return smp_send_pairing_random(smp);
 #endif /* CONFIG_BT_PERIPHERAL */
 
 	return 0;
```
Dead return `0` here if CONFIG_BT_PERIPHERAL is defined, Likely build warning/failure here.
It would be a failure if you end up on this return with no peripheral defined. 
```
#else
return BT_SMP_ERR_UNPSECIFIED
#endif /*CONFIG_BT_PERIPHERAL */
```

Vudentz replied:
I guess we should be using BT_SMP_ERR_PAIRING_NOTSUPP since we cannot pair as peripheral.

joerchan replied:
Sounds good.

carlescufi commented:
@Vudentz you have a `!fixup` commit in there
